File: implementation2/src/bot_behavior/bot_config.ts
import {
  BotInternalState,
  BotMemory,
  BotBehavior,
  WanderState,
  EscapeState,
  GetPowerupState,
  AttackState,
  HostileBot,
  CarefulBot,
  GreedyBot,
  Model,
  World,
  Player,
  GridCoords,
  Direction,
  Bomb,
  Explosion,
  BotType,
  BombOnlyDangerPolicy,
  AttackPolicy2,
  PowerupPolicy2,
  ExplosionPredictionDangerPolicy,
  AttackPolicy1,
  PowerupPolicy1,
  BotUpdateResult,
  BotConfig,
  IdleAction,
  Entity,
} from "../model";
import { Schema as S, HashSet, Match, Option } from "effect";

// FACTORY AND INIT
export const createBotConfig = (botType: BotType): BotConfig => {
  return Match.value(botType).pipe(
    Match.tag("Hostile Bot", (bot) => ({
      botType: bot,
      reevalInterval: 0.5,
      reevalChance: 0.25,
      dangerRadius: 0,
      dangerPolicy: BombOnlyDangerPolicy.make({}),
      attackPolicy: AttackPolicy2.make({}),
      attackRangeTrigger: 2,
      attackSearchRadius: 0,
      powerupPolicy: PowerupPolicy2.make({}),
      powerupChance: 0.2,
    })),

    Match.tag("Careful Bot", (bot) => ({
      botType: bot,
      reevalInterval: 0.25,
      reevalChance: 1.0,
      dangerRadius: 4,
      dangerPolicy: ExplosionPredictionDangerPolicy.make({}),
      attackPolicy: AttackPolicy1.make({ maxDistance: 3 }),
      attackRangeTrigger: 4,
      attackSearchRadius: 3,
      powerupPolicy: PowerupPolicy2.make({}),
      powerupChance: 1.0,
    })),

    Match.tag("Greedy Bot", (bot) => ({
      botType: bot,
      reevalInterval: 1.0,
      reevalChance: 1.0,
      dangerRadius: 2,
      dangerPolicy: ExplosionPredictionDangerPolicy.make({}),
      attackPolicy: AttackPolicy1.make({ maxDistance: 6 }),
      attackRangeTrigger: 3,
      attackSearchRadius: 6,
      powerupPolicy: PowerupPolicy1.make({}),
      powerupChance: 1.0,
    })),

    Match.exhaustive
  );
}

File: implementation1/bot_behavior/bot_config.py
from dataclasses import dataclass
from bot_behavior.bot import BotController
from bot_behavior.bot_policy import AttackPolicy1, AttackPolicy2, BombOnlyDangerPolicy, ExplosionPredictionDangerPolicy, PowerupPolicy1, PowerupPolicy2
from common_types import BotConfigInfo, BotControllerInfo, BotType


@dataclass(frozen=True)
class HostileBotConfig:
    bot_type = BotType.HOSTILE

    reeval_interval = 0.5
    reeval_chance = 0.25

    danger_radius = 0
    danger_check_type = BombOnlyDangerPolicy()

    attack_policy = AttackPolicy2()
    attack_range_trigger = 2
    attack_search_radius = 0

    powerup_policy = PowerupPolicy2()
    powerup_chance = 0.20

@dataclass(frozen=True)
class CarefulBotConfig:
    bot_type = BotType.CAREFUL

    reeval_interval = 0.25
    reeval_chance = 1.0

    danger_radius = 4
    danger_check_type = ExplosionPredictionDangerPolicy()

    attack_policy = AttackPolicy1(max_distance=3)
    attack_range_trigger = 4
    attack_search_radius = 3

    powerup_policy = PowerupPolicy2()
    powerup_chance = 1.0


@dataclass(frozen=True)
class GreedyBotConfig:
    bot_type = BotType.GREEDY

    reeval_interval = 1.0
    reeval_chance = 1.0

    danger_radius = 2
    danger_check_type = ExplosionPredictionDangerPolicy()

    attack_policy = AttackPolicy1(max_distance=6)
    attack_range_trigger = 3
    attack_search_radius = 6

    powerup_policy = PowerupPolicy1()
    powerup_chance = 1.0

@dataclass(frozen=True)
class ExtremeBotConfig:
    bot_type = BotType.EXTREME

    reeval_interval = 0.1
    reeval_chance = 0.1

    danger_radius = 10
    danger_check_type = ExplosionPredictionDangerPolicy()

    attack_policy = AttackPolicy2()
    attack_range_trigger = 10
    attack_search_radius = 0

    powerup_policy = PowerupPolicy1()
    powerup_chance = 1.0

class BotFactory:
    _CONFIG_MAP: dict[BotType, type[BotConfigInfo]] = {
        BotType.HOSTILE: HostileBotConfig,
        BotType.CAREFUL: CarefulBotConfig,
        BotType.GREEDY: GreedyBotConfig,
        BotType.EXTREME: ExtremeBotConfig,
        }

    @classmethod
    def make(cls, bot_type: BotType) -> BotControllerInfo:
        config = cls.create_bot_config(bot_type)
        return BotController(config)

    @classmethod
    def create_bot_config(cls, bot_type: BotType) -> BotConfigInfo:
        config_class = cls._CONFIG_MAP.get(bot_type)
        
        if config_class is None:
            raise ValueError(f"Unknown bot type: {bot_type}")
        
        return config_class()
